src/compiler/checker.ts

```diff
@@ -9702,6 +9702,41 @@ namespace ts {
             return links.type;
         }
 
+        function getWriteTypeOfSymbolWithDeferredType(symbol: Symbol): Type | undefined {
+            const links = getSymbolLinks(symbol);
+            if (!links.writeType && links.deferralWriteConstituents) {
+                Debug.assertIsDefined(links.deferralParent);
+                Debug.assertIsDefined(links.deferralConstituents);
+                links.writeType = links.deferralParent.flags & TypeFlags.Union ? getUnionType(links.deferralWriteConstituents) : getIntersectionType(links.deferralWriteConstituents);

```

Comment:
Shouldn't the write type of a union be resolved into an intersection type? Just like the way mentioned in https://github.com/microsoft/TypeScript/pull/30769

Reply:
If you come up with a code example that doesn’t work the way you expect, feel free to file an issue.

Reply:
> If you come up with a code example that doesn’t work the way you expect, feel free to file an issue.

Please have a look at the unsound behavior described in https://github.com/microsoft/TypeScript/issues/50142. [A comment](https://github.com/microsoft/TypeScript/issues/50142#issuecomment-1203042758) indicates that it should be treated using the same rule of https://github.com/microsoft/TypeScript/pull/30769.

I'm not quite sure but I guess it could have relation to this PR, where the write type of a union is not resolved into an intersection, but a union.

Reply:
FWIW there’s a second complication in play in #50142: subtype reduction.  Subtype reduction means that `NullableProp | NonNullableProp` is likely to be reduced to just the former, making #30769 a moot point.